std/protocols/permutation.asm

```diff
@@ -1,44 +1,152 @@
 use std::prover::challenge;
 use std::array::fold;
+use std::utils::unwrap_or_else;
 use std::array::len;
-use std::convert::int;
 use std::check::assert;
-use std::field::modulus;
+use std::check::panic;
+use std::field::known_field;
+use std::field::KnownField;
+use std::math::fp2::Fp2;
+use std::math::fp2::add_ext;
+use std::math::fp2::sub_ext;
+use std::math::fp2::mul_ext;
+use std::math::fp2::unpack_ext;
+use std::math::fp2::next_ext;
+use std::math::fp2::inv_ext;
+use std::math::fp2::eval_ext;
+use std::math::fp2::from_base;
 
 let is_first: col = |i| if i == 0 { 1 } else { 0 };
 
 // Get two phase-2 challenges to use in all permutation arguments.
-// Note that this assumes that globally no other challenge of these IDs is used.
-let alpha: expr = challenge(0, 1);
-let beta: expr = challenge(0, 2);
+// Note that this assumes that globally no other challenge of these IDs is used,
+// and that challenges for multiple permutation arguments are re-used.
+// We declare two components for each challenge here, in case we need to operate
+// on the extension field. If we don't, we won't end up needing it and the optimizer
+// will remove it.
+let alpha1: expr = challenge(0, 1);
+let alpha2: expr = challenge(0, 2);
 
-// Maps [x_1, x_2, ..., x_n] to alpha**(n - 1) * x_1 + alpha ** (n - 2) * x_2 + ... + x_n
-let compress_expression_array = |expr_array| fold(expr_array, 0, |acc, el| alpha * acc + el);
+let beta1: expr = challenge(0, 3);
+let beta2: expr = challenge(0, 4);
+
+let unpack_permutation_constraint: Constr -> (expr, expr[], expr, expr[]) = |permutation_constraint| match permutation_constraint {
+    Constr::Permutation(lhs_selector, lhs, rhs_selector, rhs) => (unwrap_or_else(lhs_selector, || 1), lhs, unwrap_or_else(rhs_selector, || 1), rhs),
+    _ => panic("Expected permutation constraint")
+};
+
+/// Whether we need to operate on the F_{p^2} extension field (because the current field is too small).
+let needs_extension: -> bool = || match known_field() {
+    Option::Some(KnownField::Goldilocks) => true,
```

Comment:
We could also differentiate based on the modulus, but maybe this is a bit safer.